### PR TITLE
Bit-exact GaussianBlur performance update

### PR DESCRIPTION
**Pull request by terfendail**
### This pullrequest changes
Reworked evaluation of bit-exact GaussianBlur to improve performance for generic kernel size

### CODE REVIEW

File: modules/imgproc/src/smooth.cpp [hunk 1]
```diff
@@ -2819,24 +3282,33 @@ template <typename ET, typename FT>
...
-        dst[i] = (FT::WT(src[2][i])*6 + ((FT::WT(src[1][i]) + FT::WT(src[3][i]))<<2) + FT::WT(src[0][i]) + FT::WT(src[4][i])) >> 4;
+        dst[i] = (FT::WT(src[2][i])*(uint8_t)6 + ((FT::WT(src[1][i]) + FT::WT(src[3][i]))<<2) + FT::WT(src[0][i]) + FT::WT(src[4][i])) >> 4;
 }
 template <>
 void vlineSmooth5N14641<uint8_t, ufixedpoint16>(const ufixedpoint16* const * src, const ufixedpoint16*, int, uint8_t* dst, int len)
 {
     int i = 0;
     v_uint32x4 v_6 = v_setall_u32(6);
-    for (; i < len - 7; i += 8)
+    for (; i < len - 15; i += 16)
```
Comment (alalek) [inline modules/imgproc/src/smooth.cpp]: Current OpenCV pattern for such loops is:
```
for (; i <= len - <vec_width>; i += <vec_width>)
{
    ...
}
```

Using of different constants should be avoided (16/15)

File: modules/imgproc/src/smooth.cpp [hunk 2]
```diff
@@ -2770,47 +3185,95 @@ void vlineSmooth5N(const FT* const * src, const FT* m, int, ET* dst, int len)
...
+        v_int32x4 v_res7 = v_dotprod(v_tmp1, v_mul01);
+
+        v_src00 = v_load((int16_t*)(src[2]) + i);
+        v_src01 = v_load((int16_t*)(src[2]) + i + 8);
+        v_src02 = v_load((int16_t*)(src[2]) + i + 16);
+        v_src03 = v_load((int16_t*)(src[2]) + i + 24);
+        v_src10 = v_load((int16_t*)(src[3]) + i);
+        v_src11 = v_load((int16_t*)(src[3]) + i + 8);
+        v_src12 = v_load((int16_t*)(src[3]) + i + 16);
+        v_src13 = v_load((int16_t*)(src[3]) + i + 24);
```
Comment (alalek) [inline modules/imgproc/src/smooth.cpp]: SSE/AVX instruction set has 16 vector registers only.

`v_res{0, 7}` (8) and `v_src*` (8) + external `v_mul*` + other exceeds registers pool.

Are you sure that there is no register spilling into stack?

**Comment by vpisarev** [comment]
:+1:

File: modules/imgproc/src/fixedpoint.inl.hpp [hunk 3]
```diff
@@ -61,23 +67,29 @@ class fixedpoint64
         uint64_t val0_h = (sh2   & 0xFFFFFFFF) + (sh1_0 >> 32) + (sh1_1 >> 32) + (val0_l >> 32);
         val0_l &= 0xFFFFFFFF;
 
+        if ( (sh2 >> 32) || (val0_h >> ressign ? 32 : 31) )
+            return (ressign ? ~(int64_t)0x7FFFFFFFFFFFFFFF : (int64_t)0x7FFFFFFFFFFFFFFF);
```
Comment (alalek) [inline modules/imgproc/src/fixedpoint.inl.hpp]: Created issue #11330

**Pull request merged**